File: web/src/crypto/keyring.ts
import { importKey } from './aead';
import { b64ToBytes, type B64 } from './bytes';
import { sealInfo } from './envelope';
import { openGroupKey } from './group';
import type { Identity } from './identity';
import { open } from './sealedbox';

export interface KeyGrantDto {
  scope_id: number;
  /** Names the scope inside the sealed box, so a grant cannot be replayed elsewhere. */
  scope_client_id: string;
  key_version: number;
  subject_type: 'user' | 'group' | 'invite' | 'share_link';
  subject_id: number;
  wrapped_key: B64;
  nonce: B64;
  wrap_algorithm: string;
}

/** The caller's own copy of a group's private key, as the server returns it. */
export interface GroupKeyDto {
  group_id: number;
  group_client_id: string;
  key_version: number;
  wrapped_key: B64;
  nonce: B64;
}

const SUPPORTED_ALGORITHM = 'ecdh-p256-hkdf-a256gcm';

/**
 * Every content key the viewer can open, indexed by scope and version.
 *
 * Old versions are kept rather than replaced: revisions and trashed items stay encrypted
 * under whatever version was current when they were written, so dropping them would make
 * history unreadable after the first key rotation.
 *
 * A missing key is the normal case, not an error — it is exactly what a node the viewer
 * has no access to looks like.
 */
export class ScopeKeyring {
  private readonly keys = new Map<string, CryptoKey>();
  private readonly scopes = new Set<number>();

  /**
   * Builds the keyring from what the server hands back.
   *
   * A scope key sealed to a group is opened with the group's private key rather than the
   * reader's own — which is the point of a group: one seal per scope, not one per member.
   * Those private keys arrive separately, each sealed to the reader, so a group they do
   * not belong to yields nothing.
   */
  static async fromGrants(
    grants: KeyGrantDto[],
    identity: Identity,
    groupKeys: GroupKeyDto[] = [],
  ): Promise<ScopeKeyring> {
    const keyring = new ScopeKeyring();
    const groups = new Map<number, CryptoKey>();

    for (const key of groupKeys) {
      try {
        groups.set(
          key.group_id,
          await openGroupKey(
            identity,
            { blob: b64ToBytes(key.wrapped_key), nonce: b64ToBytes(key.nonce) },
            key.group_client_id,
            key.key_version,
          ),
        );
      } catch {
        // A copy this device cannot open belongs to an older version of the group, or to
        // a format this client no longer speaks. Neither should cost the whole vault.
      }
    }

    for (const grant of grants) {
      if (grant.wrap_algorithm !== SUPPORTED_ALGORITHM) continue;

      const opener =
        grant.subject_type === 'user'
          ? identity.sealPrivate
          : grant.subject_type === 'group'
            ? groups.get(grant.subject_id)
            : undefined;

      if (!opener) continue;

      try {
        const raw = await open(
          opener,
          { blob: b64ToBytes(grant.wrapped_key), nonce: b64ToBytes(grant.nonce) },
          sealInfo(grant.scope_client_id, grant.key_version),
        );

        keyring.add(grant.scope_id, grant.key_version, await importKey(raw));
      } catch {
        // A grant that will not open is a grant for somebody else, or one written by a
        // client that has since changed format. Neither is worth failing the whole vault.
      }
    }

    return keyring;
  }

  add(scopeId: number, version: number, key: CryptoKey): void {
    this.keys.set(ScopeKeyring.at(scopeId, version), key);
    this.scopes.add(scopeId);
  }

  get(scopeId: number, version: number): CryptoKey | undefined {
    return this.keys.get(ScopeKeyring.at(scopeId, version));
  }

  /** Drives which drop targets the tree offers: a scope with no key cannot receive a move. */
  has(scopeId: number): boolean {
    return this.scopes.has(scopeId);
  }

  get size(): number {
    return this.keys.size;
  }

  private static at(scopeId: number, version: number): string {
    return `${scopeId}:${version}`;
  }
}

File: internal/mcp/keyring.go
package mcp

import (
	"shelf/internal/envelope"
	"shelf/internal/vault"
)

// supportedAlgorithm is the only wrapping this side reads. A grant written under anything
// else is skipped rather than guessed at.
const supportedAlgorithm = "ecdh-p256-hkdf-a256gcm"

// Keyring holds the scope keys a connector can open, by scope and version.
//
// Old versions are kept alongside current ones for the same reason the browser keeps them:
// the trash and the revision history are sealed under the key that was current when they
// were written, and dropping it would make them unreadable rather than merely stale.
type Keyring struct {
	keys map[slot][]byte
}

type slot struct {
	scopeID int64
	version int32
}

// NewKeyring opens every grant it can and passes over the rest.
//
// A grant that will not open is the ordinary case, not a failure: a folder given its own key
// is simply outside what this connector may see. Refusing to build the ring would turn a
// vault with one private folder into a vault the connector cannot read at all.
func NewKeyring(identity *envelope.Identity, grants []vault.KeyGrant) *Keyring {
	ring := &Keyring{keys: make(map[slot][]byte, len(grants))}

	for _, grant := range grants {
		if grant.Algorithm != "" && grant.Algorithm != supportedAlgorithm {
			continue
		}

		key, err := envelope.Open(
			identity.Seal,
			envelope.Box{Blob: grant.WrappedKey, Nonce: grant.Nonce},
			envelope.SealInfo(grant.ScopeClientID, grant.KeyVersion),
		)
		if err != nil || len(key) != envelope.KeyLength {
			continue
		}

		ring.keys[slot{scopeID: grant.ScopeID, version: grant.KeyVersion}] = key
	}

	return ring
}

// Get returns the key for one slot, or nil when the connector holds none. A nil key is a
// state the callers render as locked rather than an error they abort on.
func (k *Keyring) Get(scopeID int64, version int32) []byte {
	return k.keys[slot{scopeID: scopeID, version: version}]
}

// Has reports whether any version of a scope is readable, which is what decides whether a
// subtree is offered at all.
func (k *Keyring) Has(scopeID int64) bool {
	for at := range k.keys {
		if at.scopeID == scopeID {
			return true
		}
	}

	return false
}

// Len is the number of slots opened, for logging that says how much of a vault is reachable
// without saying anything about what is in it.
func (k *Keyring) Len() int { return len(k.keys) }
